doc: move Thingy:53 getting started guide (#7110)
nrfconnect/sdk-nrf: doc/nrf/ug_thingy53.rst

.. _ug_thingy53:

Developing with Thingy:53
#########################

.. contents::
   :local:
   :depth: 2

The Nordic Thingy:53 is a battery-operated prototyping platform for IoT Systems.

> The Nordic Thingy:53 integrates the nRF5340 SoC that support Bluetooth® Low Energy, IEEE 802.15.4 based protocols and Near Field Communication (NFC).

Author: Fixed.

.. _thingy53_build_pgm_targets:

Build Targets
=============

The build targets of interest for Thingy:53 in the |NCS| are listed on the table below.

+--------------------------------+----------------------------------------------------------+
|Component                       |  Build target                                            |
+================================+==========================================================+
| nRF5340 SoC - Application core |``thingy53_nrf5340_cpuapp`` for the secure domain         |
|                                |                                                          |
|                                |``thingy53_nrf5340_cpuapp_ns`` for the non-secure domain  |
+--------------------------------+----------------------------------------------------------+
| nRF5340 SoC - Network core     |``thingy53_nrf5340_cpunet``                               |
+--------------------------------+----------------------------------------------------------+

.. note::
   The |NCS| samples and applications that are compatible with the Nordic Thingy:53 follow the :ref:`thingy53_app_guide`.

>    The applicaton guide defines a consistent partition map and booloader configuration to allow serial and OTA firmware updates.

Reviewer: ```suggestion
   The application guide defines a consistent partition map and bootloader configuration to allow serial and OTA firmware updates.
```

+--------------------------------+----------------------------------------------------------+
|Component                       |  Build target                                            |
+================================+==========================================================+
| nRF5340 SoC - Application core |``thingy53_nrf5340_cpuapp`` for the secure domain         |
|                                |                                                          |
|                                |``thingy53_nrf5340_cpuapp_ns`` for the non-secure domain  |
+--------------------------------+----------------------------------------------------------+
| nRF5340 SoC - Network core     |``thingy53_nrf5340_cpunet``                               |
+--------------------------------+----------------------------------------------------------+

.. note::
   The |NCS| samples and applications that are compatible with the Nordic Thingy:53 follow the :ref:`thingy53_app_guide`.
   The application guide defines a consistent partition map and bootloader configuration to allow serial and OTA firmware updates.

The |NCS| by default uses :ref:`ug_multi_image` for Thingy:53.

> Because of this, when you choose ``thingy53_nrf5340_cpuapp`` when building a sample or application that is compatible with Thingy:53, you will generate firmware for both the application core and network core:

Reviewer 2: ``thingy53_nrf5340_cpuapp`` or  ``thingy53_nrf5340_cpuapp_ns``

| :file:`dfu_application.zip`     | MCUboot compatible update images for both cores | Using nRF Programmer for Android and iOS, or nRF Connect Programmer |
+---------------------------------+-------------------------------------------------+---------------------------------------------------------------------+

See the following sections for details regarding building and programming the firmware for Thingy:53 in various environments.
If your Thingy:53 is already programmed with a Thingy:53-compatible sample or application, you can also use the MCUboot bootloader to update the firmware after you finish building.
See :ref:`thingy53_app_update` for more detailed information about updating firmware on Thingy:53.

.. _thingy53_build_pgm_vscode:

Building and programming using Visual Studio Code
=================================================

Complete the following steps after installing the `nRF Connect for Visual Studio Code`_ extension in Microsoft Visual Studio Code:

1. Open Visual Studio Code.
#. Click the :guilabel:`nRF Connect for VS Code` icon in the left navigation bar.

>    The :guilabel:`Welcome` page opens.

Reviewer: The welcome page is a different view. The button opens the nRF Connect For VS Code sidebar. Could skip this line altogether without losing anyone.

   Click the :guilabel:`Add Build Configuration` icon that appears on the right-hand side, or click the text :guilabel:`No build configurations. Click to create one`.

   This opens the :guilabel:`Add Build Configuration` window in a new tab.
#. Select the ``thingy53_nrf5340_cpuapp`` as the target board and click :guilabel:`Build Configuration`, which generates the configuration file and triggers the build process.

   If you see the :guilabel:`Generate Configuration` button instead of the :guilabel:`Build Configuration` button, select the :guilabel:`Build after generating configuration` checkbox.

   .. figure:: images/ug_thingy53_vscode_genconfig1.png
      :alt: nRF Connect for VS Code, Generate Configuration window

      The :guilabel:`Generate Configuration` window, with ``thingy53_nrf5340_cpuapp`` as the board.

   When the build configuration and the build are complete, an :guilabel:`Actions` panel appears in the lower left corner of Visual Studio Code.
   In this panel, you can trigger the build process, program the built sample, or start a debug session.

   .. figure:: images/ug_thingy53_vscode_genconfig2.png
      :alt: nRF Connect for VS Code, Actions panel

>       The :guilabel:`Actions` panel is open in the lower left corner

Author: Quite possibly. That's something to work out during the doc improvement phase, I'd say.

| :file:`dfu_application.zip`     | MCUboot compatible update images for both cores | Using nRF Programmer for Android and iOS, or nRF Connect Programmer |
+---------------------------------+-------------------------------------------------+---------------------------------------------------------------------+

See the following sections for details regarding building and programming the firmware for Thingy:53 in various environments.
If your Thingy:53 is already programmed with a Thingy:53-compatible sample or application, you can also use the MCUboot bootloader to update the firmware after you finish building.
See :ref:`thingy53_app_update` for more detailed information about updating firmware on Thingy:53.

.. _thingy53_build_pgm_vscode:

Building and programming using Visual Studio Code
=================================================

Complete the following steps after installing the `nRF Connect for Visual Studio Code`_ extension in Microsoft Visual Studio Code:

1. Open Visual Studio Code.
#. Click the :guilabel:`nRF Connect for VS Code` icon in the left navigation bar and then click :guilabel:`Open`.

   The :guilabel:`Welcome` panel opens.

>    .. figure:: images/ug_thingy53_vscode_welcome.png

Author: Taking it out. Should be clear without the screenshot.

* Intel Hex file (:file:`merged_domains.hex`) - Used with an external debug probe.
  The file contains bootloaders and applications for both cores.
* Binary files (:file:`app_update.bin`, :file:`net_core_app_update.bin`), containing signed application firmwares for the application and network core, respectively.
  For convenience, the binary files are bundled in :file:`dfu_application.zip`, together with a manifest that describes them.

>   The binary files or the combined zip archive can be used to update application firmware  for both cores, with either MCUboot serial recovery or OTA DFU using Bluetooth LE.

Author: Fixed.

| :file:`app_update.bin`          | MCUboot compatible application core update      | Using the built-in bootloader and nRF Connect Programmer            |
+---------------------------------+-------------------------------------------------+---------------------------------------------------------------------+
| :file:`net_core_app_update.bin` | MCUboot compatible network core update          | Using the built-in bootloader and nRF Connect Programmer            |
+---------------------------------+-------------------------------------------------+---------------------------------------------------------------------+
| :file:`dfu_application.zip`     | MCUboot compatible update images for both cores | Using nRF Programmer for Android and iOS, or nRF Connect Programmer |
+---------------------------------+-------------------------------------------------+---------------------------------------------------------------------+

See the following sections for details regarding building and programming the firmware for Thingy:53 in various environments.

> If your Thingy:53 is already programmed with Thingy:53-compatible sample or application, then you can also use the MCUboot bootloader to update the firmware after you finish building.

Author: Fixed.

See :ref:`thingy53_app_update` for more detailed information about updating firmware on Thingy:53.

.. _thingy53_build_pgm_vscode:

Building and programming using Visual Studio Code
=================================================

Complete the following steps after installing the `nRF Connect for Visual Studio Code`_ extension in Microsoft Visual Studio Code:

> 1. Open Visual Studio code.

Author: Fixed.

> #. Click on the :guilabel:`nRF Connect for VS Code` icon in the left navigation bar and then click :guilabel:`Open`.

Author: Fixed.

   The :guilabel:`Welcome` panel opens.

   .. figure:: images/ug_thingy53_vscode_welcome.png
      :alt: nRF Connect for VS Code welcome panel

      The welcome panel for nRF Connect for VS Code.

> #. Click on :guilabel:`Add an existing application to workspace...` in the :guilabel:`Welcome` panel.

Author: Fixed.

#. In the prompt, navigate to the folder containing the sample you want to build, such as :file:`nrf/samples/bluetooth/peripheral_lbs`.

>    You will now see the selected application in the :guilabel:`Applications` panel in the lower left corner.

Author: Fixed.

   .. note::

     The sample folder must contain a :file:`prj.conf` file.

   .. figure:: images/ug_thingy53_vscode_apppanel.png
      :alt: nRF Connect for VS Code, application shown in Applications panel

>       The peripheral_lbs application is now available in the :guilabel:`Applications` panel.

Author: Also made it clearer that `peripheral_lbs` is an example.

> #. Click on the :guilabel:`Add Build Configuration` button in the :guilabel:`Applications` panel, or click on the text stating :guilabel:`No build configurations. Click to create one`.

Author: Fixed.

   This opens the :guilabel:`Add Build Configuration` window in a new tab.
#. Select the ``thingy53_nrf5340_cpuapp`` as the target board and click :guilabel:`Build Configuration`, which generates the configuration file and triggers the build process.

>    If you see the :guilabel:`Generate Configuration` button instead of the :guilabel:`Build Configuration` button, enable the :guilabel:`Build after generating configuration` checkbox.

Author: Fixed.

   .. figure:: images/ug_thingy53_vscode_genconfig1.png
      :alt: nRF Connect for VS Code, Generate Configuration window
      :width: 80%

      The :guilabel:`Generate Configuration` window, with ``thingy53_nrf5340_cpuapp`` as the board.

   When the build configuration and the build are complete, an :guilabel:`Actions` panel appears in the lower left corner of Visual Studio Code.

>    In this panel, you can trigger the build process, program the built sample or start a debug session.

Author: Fixed.

   .. figure:: images/ug_thingy53_vscode_genconfig2.png
      :alt: nRF Connect for VS Code, Actions panel
      :width: 80%

      The :guilabel:`Actions` panel is open in the lower left corner

#. Program the sample or application:

   a. Connect the Nordic Thingy:53 to the debug out port on a 10-pin external debug probe, for example nRF5340 DK, using a 10-pin JTAG cable.
   #. Connect the external debug probe to the PC using a USB cable.
   #. Make sure that the Nordic Thingy:53 and the external debug probe are powered on.
   #. Click :guilabel:`Flash` in the :guilabel:`Actions` panel.

.. _thingy53_build_pgm_segger:

> Building and programming using SEGGER Embedded Studio

Author: Agreed.

   .. parsed-literal::
      :class: highlight

      west build -b *build_target* -d *destination_directory_name*

>    The build target should be ``thingy53_nrf5340_cpuapp`` or ``thingy53_nrf5340_cpuapp_ns`` when building samples for the application core and ``thingy53_nrf5340_cpunet`` when building for the network core.

Reviewer 2: Should we state that in most cases we use ``thingy53_nrf5340_cpuapp`` or ``thingy53_nrf5340_cpuapp_ns`` and cpunet firmware is built and flashed as a child image in the NCS ?

Author: That is already mentioned once on Lin 63, not sure whether or not it needs to be repeated here.

Reviewer 2: In general we do not use `thingy53_nrf5340_cpunet` as a target board in Thingy:53 compatible samples (the netcore firmware is built by multiimage build system). Still the `thingy53_nrf5340_cpunet` board could be used e.g. with external programmer. Let's leave it assuming that user is aware what he's doing :)

#. Program the sample or application:

   a. Connect the Nordic Thingy:53 to the debug out port on a 10-pin external debug probe, for example nRF5340 DK (Development Kit), using a 10-pin JTAG cable.
   #. Connect the external debug probe to the PC using a USB cable.
   #. Make sure that the Thingy:53 and the external debug probe are powered on.

>    #. In SES, select :guilabel:`Target` > :guilabel:`Connect J-Link`.

Author: Fixed here and Line 174.

   #. Select :guilabel:`Target` > :guilabel:`Download zephyr/merged.hex` to program the sample or application onto Thingy:53.

>    #. The device will reset and run the programmed sample or application.

Author: Fixed.

.. caution::

>    Programming :file:`merged.hex` file updates only the application core.

Author: Fixed.

>    If you also need to update the network core, you must follow additional steps described in :ref:`ug_nrf5340`.

Author: Follow.

   Configurations that enable the other features are not mandatory, but you can enable them if needed.

   All applications compatible with the Nordic Thingy:53 in the |NCS| also support the additional features.
   Refer to the source code of the samples and applications for examples of implementation.

.. _thingy53_app_partition_manager_config:

Partition manager configuration
===============================

The samples compatible with the Nordic Thingy:53 use :ref:`partition_manager` to define memory partitions.
To ensure that partition layout does not change between builds, the sample must use a static partition layout that is the same for all compatible samples in the |NCS|.
See :file:`pm_static_thingy53_nrf5340_cpuapp.yml` in a compatible sample for the configuration used.
The memory layout must stay consistent, so that MCUboot can perform proper image updates and clean up the settings storage partition.

The PCD SRAM partition is locked by the MCUboot bootloader to prevent the application from modifying the network core firmware.
Trying to access data on this partition results in an ARM fault.

> It is also necessary to add a flash controller overlay for the application core to allow data to be loaded from the external flash memory to the network core using the application core RAM.

Reviewer 2: should we mention that this is requirement for the MCUboot ?
The image swap operation is performed by the MCUboot bootloader.

This is done because the network core cannot access the external flash memory directly.
See :file:`thingy53_nrf5340_cpuapp.overlay` in a compatible sample for how this is done.

.. _thingy53_app_mcuboot_bootloader:

MCUboot bootloader
==================

> Each sample must enable the MCUboot bootloader and use the same configuration of the bootloader as all of the Nordic Thingy:53-compatible samples in the |NCS|.

Reviewer 2: We could mention that MCUboot configuration for Thingy:53 is added as Thingy:53's board specific configuration overlay defined in the MCUboot repository (to help users locate it quickly)

Author: There's an overlay file that users need to get from the repository in addition to using the Kconfig options presented in this section? Or there is an overlay file that can be used to set these instead of doing it manually?

Reviewer 2: The MCUboot repository contains default configuration for Thingy:53 which is taken into account when building MCUboot bootloader for Thingy:53 in the NCS (for both secure and not secure variant). The configuration is added automatically when MCUboot is built for Thingy:53, user does not to do anything manually.

Author: Would adding the following sentence as a third sentence in the first paragraph convey this:
`When using |NCS| to build the MCUboot bootloader for the Thingy:53, the configuration is applied automatically from the MCUboot repository.`

This ensures that the sample includes the MCUboot bootloader and that an MCUboot-compatible image is generated when the sample is built.

The MCUboot bootloader supports serial recovery and a custom command to erase the settings storage partition.
Erasing the settings partition is needed to ensure that an application is not booted with incompatible content loaded from the settings partition.

Use the following Kconfig options to enable the MCUboot bootloader:

* :kconfig:option:`CONFIG_BOOTLOADER_MCUBOOT`
* :kconfig:option:`CONFIG_IMG_MANAGER`
* :kconfig:option:`CONFIG_MCUBOOT_IMG_MANAGER`
* :kconfig:option:`CONFIG_IMG_ERASE_PROGRESSIVELY`

> In addition, set the :kconfig:option:`CONFIG_UPDATEABLE_IMAGE_NUMBER` option to ``2`` and :kconfig:option:`CONFIG_MCUBOOT_IMAGE_VERSION` to ``"1.0.0+0"``.

Reviewer 2: The image version could have any number I think. I would not refer to specific one here

   Configurations that enable the other features are not mandatory, but you can enable them if needed.

   All applications compatible with the Nordic Thingy:53 in the |NCS| also support the additional features.
   Refer to the source code of the samples and applications for examples of implementation.

.. _thingy53_app_partition_manager_config:

Partition manager configuration
===============================

The samples compatible with the Nordic Thingy:53 use :ref:`partition_manager` to define memory partitions.
To ensure that partition layout does not change between builds, the sample must use a static partition layout that is consistent all compatible samples in the |NCS|.
See :file:`pm_static_thingy53_nrf5340_cpuapp.yml` in a compatible sample for the configuration used.
The memory layout must stay consistent, so that MCUboot can perform proper image updates and clean up the settings storage partition.

The PCD SRAM partition is locked by the MCUboot bootloader to prevent the application from modifying the network core firmware.
Trying to access data on this partition results in an ARM fault.

> You must add a flash controller overlay for the application core to allow the MCUboot bootloader load data from the external flash memory to the network core using the application core RAM.

Reviewer 2: You are mixing 2 things here:
The overlay is added to both MCUboot and to the application firmware.

1. MCUboot
The overlay with flash sim is added automatically in the background by the `nrf/modules/mcuboot/CMakeLists.txt`.  The overlay is needed for the described net core image updated.
2. Application
The overlay is added by the `thingy53_nrf5340_cpuapp.overlay` file from app configuration directory, but for different reason:
`The partition must be enabled by the application configuration to prevent bus faults when mcumgr tries to access simulated partition with network core image data.`

Reviewer 2: Can we align it a bit?

Author: ```suggestion
Both the MCUboot bootloader and the application core need a flash controller overlay.
For the MCUboot bootloader this is needed for the network core image update, and the overlay is automatically applied.
For the application core this is needed to prevent bus faults when MCUmanager tries to access the simulated partition that contains the network core image data.
See :file:`thingy53_nrf5340_cpuapp.overlay` in a compatible sample for how this is done for the application core.
```
Would this work as a replacement of the current Lines 226 - 228?

Reviewer 2: I think so. Thanks

Reviewer 2: `Both the MCUboot bootloader and the application need a flash controller overlay.`

Reviewer 2: The MCUboot bootloader and user application both run on the application core ;)

Author: Right. Changed all references from "application core" to "application" in this bit.

USB
===

An application compatible with the Nordic Thingy:53 is expected to enable USB CDC ACM as backend for logging.
The logs are provided using USB CDC ACM to allow accessing them without additional hardware.

Most of the applications and samples compatible with the Nordic Thingy:53 use only a single instance of USB CDC ACM that works as the logger's backend.
No other USB classes are used.
These samples can share a common USB product name, vendor ID, and product ID.
If a sample supports additional USB classes or more than one instance of USB CDC ACM, it must use a dedicated product name, vendor ID, and product ID.

To enable the USB device stack and the CDC ACM class:

* Use the options :kconfig:option:`CONFIG_USB_DEVICE_STACK` and :kconfig:option:`CONFIG_USB_CDC_ACM`.
* Set the default log level to Error with the :kconfig:option:`CONFIG_USB_CDC_ACM_LOG_LEVEL_ERR` and :kconfig:option:`CONFIG_USB_DEVICE_LOG_LEVEL_ERR` Kconfig options.
* Enable the UART log backend with the :kconfig:option:`CONFIG_LOG_BACKEND_UART` Kconfig option and disable SEGGER RTT with both the :kconfig:option:`CONFIG_LOG_BACKEND_RTT` and :kconfig:option:`CONFIG_USE_SEGGER_RTT` Kconfig options.

> See :file:`thingy53_nrf5340cpuapp.conf` in a compatible sample for how this is done.

Reviewer 2: `thingy53_nrf5340_cpuapp.conf`

> In addition, you have to call the :c:func:`usb_enable()` function from your application to initialize the USB stack.

Reviewer 2: `usb_enable` (without brackets) ?

See the :file:`thingy53.c` source file in a compatible sample for how this is done.

.. _thingy53_app_antenna:

Antenna selection
=================

The Nordic Thingy:53 has an RF front-end with two 2.4 GHz antennas:

* **ANT1** is connected to the nRF5340 through the nRF21540 RF FEM and supports TX gain of up to +20 dBm.
* **ANT2** is connected to the nRF5340 through the RF switch and supports TX output power of up to +3 dBm.

.. figure:: images/thingy53_antenna_connections.svg
   :alt: Nordic Thingy:53 - Antenna connections

   Nordic Thingy:53 - Antenna connections

> The samples in the |NCS| that are compatible with the Nordic Thingy:53 use **ANT1** by default, with the nRF21540 gain set to +10 dBM.

Author: These have been there a while. Thanks for catching them.

You can configure the TX gain with the :kconfig:option:`CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB` Kconfig option to select between +10 dBM or +20 dBM gain.
To use the **ANT2** antenna, disable the :kconfig:option:`CONFIG_MPSL_FEM` Kconfig option in the network core's child image configuration.

.. note::

>    Transmitting with TX output power above +10 dBM is not permitted in some regions.

Reviewer 2: `dBm`

See :file:`thing53_nrf5340_cpuapp.conf` in a compatible sample for how this is done.

In addition, the application must call the functions :c:func:`os_mgmt_register_group` and :c:func:`img_mgmt_register_group` to register the image and OS management command handlers.
The application must also call :c:func:`smp_bt_register` to register the SMP Bluetooth Service.

See the :file:`thingy53.c` source file in a compatible sample for how this is done.

.. _thingy53_app_external_flash:

External flash
==============

During a FOTA update, there might not be enough space available in internal flash storage to store the existing application and network core images as well as the incoming images, so the incoming images must be stored in external flash storage.
This means that a sample compatible with the Nordic Thingy:53 must enable the external flash using the QSPI driver.
Enable the option :kconfig:option:`CONFIG_NORDIC_QSPI_NOR`, and set :kconfig:option:`CONFIG_NORDIC_QSPI_NOR_FLASH_LAYOUT_PAGE_SIZE` to ``4096`` and :kconfig:option:`CONFIG_NORDIC_QSPI_NOR_STACK_WRITE_BUFFER_SIZE` to ``16``.

> To ensure correct operation of the flash simulator, you must also enable :kconfig:option:`CONFIG_FLASH_NOP_DEVICE`. This prevents bus faults that would happen from the MCUmanager trying to access the simulated partition with network core image data.

Reviewer 2: One sentence per line ?

Also, disable encryption and authentication requirement for the Bluetooth SMP transport using :kconfig:option:`CONFIG_MCUMGR_SMP_BT_AUTHEN`.

You need to enable the MCUmanager handlers for image and OS management.
Use the options :kconfig:option:`CONFIG_MCUMGR_CMD_IMG_MGMT` and :kconfig:option:`CONFIG_MCUMGR_CMD_OS_MGMT`.

You also have to enable the MCUmanager processing of basic Zephyr group commands and storage erase commands.
Use the options :kconfig:option:`CONFIG_MCUMGR_GRP_ZEPHYR_BASIC` and :kconfig:option:`CONFIG_MCUMGR_GRP_BASIC_CMD_STORAGE_ERASE`.

See :file:`thing53_nrf5340_cpuapp.conf` in a compatible sample for how this is done.

In addition, the application must call the functions :c:func:`os_mgmt_register_group()` and :c:func:`img_mgmt_register_group()` to register the image and OS management command handlers.
The application must also call :c:func:`smp_bt_register()` to register the SMP Bluetooth Service.

See the :file:`thingy53.c` source file in a compatible sample for how this is done.

.. _thingy53_app_external_flash:

External flash
==============

> During a FOTA update, there might not be enough available flash storage to store the existing application and network core images, so the incoming images must be stored in external flash storage.

Author: Used "as well as" instead of "in addition to".

Updating firmware image
=======================

> The firmware on the Nordic Thingy:53 can be programmed using an external debug probe and 10-pin JTAG cable, as described in the sections above for building in the different environments.

Author: Fixed sentence, added links.

You can also update applications running on both the network and application core using the built-in MCUboot bootloader and `nRF Connect Programmer`_ or the `nRF Programmer`_ app for Android and iOS.

> Prebuilt application images can also be updated that way.

Reviewer 3: ```suggestion
You can update also the prebuilt application images that way.
```

Author: Fixed.

> You can program the firmware on the Nordic Thingy:53 using an external debug probe and 10-pin JTAG cable, as described in :ref:`thingy53_building_pgming`, using either :ref:`Visual Studio Code <thingy53_build_pgm_vscode>`, :ref:`SEGGER Embedded Studio <thingy53_build_pgm_segger>`, or :ref:` command line <thingy53_build_pgm_command_line>`.

Author: Absolutely at this point.

See :ref:`thingy53_gs_updating_firmware` for details about updating firmware image.

.. _thingy53_app_update_debug:

Firmware update using external debug probe
------------------------------------------

If you are using an external debug probe, such as the nRF5340 DK, or any J-Link device supporting ARM Cortex-M33, you do not have to use applications that follow the :ref:`thingy53_app_guide`.
In that case, you can program the Thingy:53 in a similar way as nRF5340 DK.
See :ref:`ug_nrf5340` for details.

.. note::

>    You will need to program a sample or application compatible with the Nordic :ref:`thingy53_app_guide` to bring back serial recovery and DFU OTA.

Reviewer 3: ```suggestion
   You need to program a sample or application compatible with the Nordic :ref:`thingy53_app_guide` to bring back serial recovery and DFU OTA.
```

Author: Fixed.

.. _thingy53_app_update_mcuboot:

Firmware update using MCUboot bootloader
----------------------------------------

> The Thingy:53-compatible samples and applications use MCUboot bootloader that can be used to update firmware.

Author: Fixed.

> The Thingy:53-compatible samples and applications use MCUboot bootloader that you can use to update firmware.

Reviewer 4: Suggest revising to: The Thingy:53-compatible samples and applications include a MCUboot bootloader that you can use to update firmware.

Author: Applied.

  All precompiled images support OTA using Bluetooth.

See :ref:`thingy53_gs_updating_firmware` for the detailed procedures on how to program a Thingy:53 using `nRF Connect Programmer`_ or the nRF Programmer for Android or iOS.

.. _thingy53_app_guide:

Thingy:53 application guide
***************************

The Nordic Thingy:53 does not have a built-in J-Link debug IC.
Because of that, the samples and applications that are compatible with the Nordic Thingy:53 by default include MCUboot bootloader with serial recovery support.
You can update applications compatible with the Nordic Thingy:53 using the DFU functionality with either `nRF Connect Programmer`_ or the `nRF Programmer`_ mobile application.

Enabling DFU and MCUboot requires consistency in configuration of the samples and applications.
For a sample or application to be compatible with Thingy:53, the application must comply with the configurations described in the following sections.

.. note::
   To use your application with the Nordic Thingy:53 preprogrammed bootloader, you must only set the proper configuration for the Partition Manager and MCUboot.

>    Configurations that enable the other features are not mandatory, but they can be enabled if needed.

Reviewer 3: ```suggestion
   Configurations that enable the other features are not mandatory, but you can enable them if needed.
```

Author: Fixed.

The PCD SRAM partition is locked by the MCUboot bootloader to prevent the application from modifying the network core firmware.
Trying to access data on this partition results in an ARM fault.

It is also necessary to add a flash controller overlay for the application core to allow data to be loaded from the external flash memory to the network core using the application core RAM.
This is done because the network core cannot access the external flash memory directly.
See :file:`thingy53_nrf5340_cpuapp.overlay` in a compatible sample for how this is done.

.. _thingy53_app_mcuboot_bootloader:

MCUboot bootloader
==================

Each sample must enable the MCUboot bootloader and use the same configuration of the bootloader as all of the Nordic Thingy:53-compatible samples in the |NCS|.
This ensures that the sample includes the MCUboot bootloader and that an MCUboot-compatible image is generated when the sample is built.

The MCUboot bootloader supports serial recovery and a custom command to erase the settings storage partition.
Erasing the settings partition is needed to ensure that an application is not booted with incompatible content loaded from the settings partition.

> To enable the MCUboot bootloader, enable the following Kconfig options:

Reviewer 3: To enable..., enable does not sound good.

Maybe something like "Enable the following Kconfig options for the MCUboot:"

Author: Changed to: "Use the following Kconfig options to enable MCUboot bootloader:"

* :kconfig:option:`CONFIG_BOOTLOADER_MCUBOOT`
* :kconfig:option:`CONFIG_IMG_MANAGER`
* :kconfig:option:`CONFIG_MCUBOOT_IMG_MANAGER`
* :kconfig:option:`CONFIG_IMG_ERASE_PROGRESSIVELY`

In addition, set the :kconfig:option:`CONFIG_UPDATEABLE_IMAGE_NUMBER` option to ``2`` and :kconfig:option:`CONFIG_MCUBOOT_IMAGE_VERSION` to ``"1.0.0+0"``.

See :file:`thingy53_nrf5430_cpuapp.conf` in a compatible sample for how this is done.

.. _thingy53_app_usb:

USB
===

> An application compatible with the Nordic Thingy:53 is expected to enable USB CDC ACM as back-end for logging.

Reviewer 3: back end or back-end (or even backend)

Author: With a quick search, "backend" is the most common in our documentation.

The logs are provided using USB CDC ACM to allow accessing them without additional hardware.

Most of the applications and samples compatible with the Nordic Thingy:53 use only a single instance of USB CDC ACM that works as the logger's back-end.
No other USB classes are used.
These samples can share a common USB product name, vendor ID, and product ID.
If a sample supports additional USB classes or more than one instance of USB CDC ACM, it must use a dedicated product name, vendor ID, and product ID.

To enable the USB device stack and the CDC ACM class:

* Use the options :kconfig:option:`CONFIG_USB_DEVICE_STACK` and :kconfig:option:`CONFIG_USB_CDC_ACM`.
* Set the default log level to Error by using options :kconfig:option:`CONFIG_USB_CDC_ACM_LOG_LEVEL_ERR` and :kconfig:option:`CONFIG_USB_DEVICE_LOG_LEVEL_ERR`.

> * Enable the UART log back-end by using the :kconfig:option:`CONFIG_LOG_BACKEND_UART` option and disable SEGGER RTT by using both :kconfig:option:`CONFIG_LOG_BACKEND_RTT` and :kconfig:option:`CONFIG_LOG_BACKEND_RTT`.

Reviewer 3: Can you somehow get rid of the "...by using..." structure?

Author: replaced "by using" with "with".

See :file:`thingy53_nrf5340cpuapp.conf` in a compatible sample for how this is done.

In addition, you have to call the :c:func:`usb_enable()` function from your application to initialize the USB stack.

See the :file:`thingy53.c` source file in a compatible sample for how this is done.

.. _thingy53_app_antenna:

Antenna selection
=================

The Nordic Thingy:53 has an RF front-end with two 2.4 GHz antennas:

* **ANT1** is connected to the nRF5340 through the nRF21540 RF FEM and supports TX gain of up to +20 dBm.
* **ANT2** is connected to the nRF5340 through the RF switch and supports TX output power of up to +3 dBm.

.. figure:: images/thingy53_antenna_connections.svg
   :alt: Nordic Thingy:53 - Antenna connections

   Nordic Thingy:53 - Antenna connections

The samples in the |NCS| that are compatible with the Nordic Thingy:53 use **ANT1** by default, with the nRF21540 gain set to +10 dBM.

> The TX gain can be configured with the :kconfig:option:`CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB` Kconfig option to select between +10 dBM or +20 dBM gain.

Reviewer 3: ```suggestion
You can configure the TX gain with the :kconfig:option:`CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB` Kconfig option to select between +10 dBM or +20 dBM gain.
```

Author: Fixed.

Thingy:53 application guide
***************************

The Nordic Thingy:53 does not have a built-in J-Link debug IC.
Because of that, the samples and applications that are compatible with the Nordic Thingy:53 by default use MCUboot bootloader with serial recovery support.
Applications compatible with the Nordic Thingy:53 can be updated using DFU functionality with either `nRF Connect Programmer`_ or the `nRF Programmer`_ mobile application.

Enabling DFU and MCUboot requires consistency in configuration of the samples and applications.
For a sample or application to be compatible with Thingy:53, the application must comply with the configurations described in the following sections.

.. note::

>    To use your application with the Nordic Thingy:53 preprogrammed bootloader, you only must set proper configuration of Partition Manager and MCUboot.

Author: Fixed. Thanks, that was bothering me but I couldn't see the fix myself.

See :file:`thingy53_nrf5340_cpuapp.overlay` in a compatible sample for how this is done.

.. _thingy53_app_mcuboot_bootloader:

MCUboot bootloader
==================

Each sample must enable the MCUboot bootloader and use the same configuration of the bootloader as all of the samples in the |NCS| that are compatible with the Nordic Thingy:53.
This is needed to let MCUboot use the sample as an image update.

The MCUboot bootloader supports serial recovery and a custom command to erase the settings storage partition.
Erasing the settings partition is needed to ensure that an application is not booted with incompatible content loaded from the settings partition.

The MCUboot bootloader is enabled by using the following Kconfigs:

* :kconfig:`CONFIG_BOOTLOADER_MCUBOOT`
* :kconfig:`CONFIG_IMG_MANAGER`
* :kconfig:`CONFIG_MCUBOOT_IMG_MANAGER`
* :kconfig:`CONFIG_IMG_ERASE_PROGRESSIVELY`

In addition, :kconfig:`CONFIG_UPDATEABLE_IMAGE_NUMBER` must be set to ``2`` and :kconfig:`CONFIG_MCUBOOT_IMAGE_VERSION` must be set to ``"1.0.0+0"``.

See :file:`thingy53_nrf5430_cpuapp.conf` in a compatible sample for how this is done.

.. _thingy53_app_usb:

USB
===

> An application compatible with the Nordic Thingy:53 is expected to enable USB CDC ACM as backend for logging.

Author: Fixed.

The logs are provided using USB CDC ACM to allow accessing them without additional hardware.

> Most of the applications and samples compatible with the Nordic Thingy:53 use only a single instance of USB CDC ACM that works as logger's backend.

Reviewer 5: ```suggestion
Most of the applications and samples compatible with the Nordic Thingy:53 use only a single instance of USB CDC ACM that works as the logger's back-end.
```

Author: Fixed.

No other USB classes are used.
These samples can share a common USB product name, vendor ID, and product ID.
If a sample supports additional USB classes or more than one instance of USB CDC ACM, it must use a dedicated product name, vendor ID, and product ID.

The USB device stack and the CDC ACM class are enabled by using :kconfig:`CONFIG_USB_DEVICE_STACK` and :kconfig:`CONFIG_USB_CDC_ACM`.
The default log level should be set to Error.
This is done by enabling :kconfig:`CONFIG_USB_CDC_ACM_LOG_LEVEL_ERR` and :kconfig:`CONFIG_USB_DEVICE_LOG_LEVEL_ERR`.

> Also, the UART log backend must be enabled by using :kconfig:`CONFIG_LOG_BACKEND_UART`, and SEGGER RTT must be disabled by using both :kconfig:`CONFIG_LOG_BACKEND_RTT` and :kconfig:`CONFIG_LOG_BACKEND_RTT`.

Reviewer 5: ```suggestion
Also, the UART log back-end must be enabled by using :kconfig:`CONFIG_LOG_BACKEND_UART`, and SEGGER RTT must be disabled by using both :kconfig:`CONFIG_LOG_BACKEND_RTT` and :kconfig:`CONFIG_LOG_BACKEND_RTT`.
```

Author: Fixed.

The Nordic Thingy:53 has an RF front-end with two 2.4 GHz antennas:

* **ANT1** is connected to the nRF5340 through the nRF21540 RF FEM and supports TX gain of up to +20 dBm.
* **ANT2** is connected to the nRF5340 through the RF switch and supports TX output power of up to +3 dBm.

.. figure:: images/thingy53_antenna_connections.png
   :alt: Nordic Thingy:53 - Antenna connections

   Nordic Thingy:53 - Antenna connections

The samples in the |NCS| that are compatible with the Nordic Thingy:53 use **ANT1** by default, with the nRF21540 gain set to +10 dBM.
The TX gain can be configured with the :kconfig:`CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB` Kconfig option to select between +10 dBM or +20 dBM gain.
To use the **ANT2** antenna, disable the :kconfig:`CONFIG_MPSL_FEM` Kconfig option.

.. note::
   Transmitting wth TX output power above +10 dBM is not permitted in some regions.
   See the `Nordic Thingy:53 Hardware`_ documentation for the applicable regulations in your region before changing the configuration.

.. _thingy53_app_fota_smp:

FOTA and SMP
============

FOTA updates on the Nordic Thingy:53 are supported using MCUmanager's Simple Management Protocol (SMP) over Bluetooth.

> The application acts as a GATT server and allows the cibbected Bluetooth Central to perform the firmware update for both application and network core.

Reviewer 5: ```suggestion
The application acts as a GATT server and allows the connected Bluetooth Central to perform the firmware update for both the application and network core.
```

Author: Fixed. That's an embarrassing type.

To do this, use :kconfig:`CONFIG_MCUMGR_CMD_IMG_MGMT` and :kconfig:`CONFIG_MCUMGR_CMD_OS_MGMT`.

The MCUmanager processing of basic Zephyr group commands and storage erase commands must also be enabled.
This is done using both :kconfig:`CONFIG_MCUMGR_GRP_ZEPHYR_BASIC` and :kconfig:`CONFIG_MCUMGR_GRP_BASIC_CMD_STORAGE_ERASE`.

See :file:`thing53_nrf5340_cpuapp.conf` in a compatible sample for how this is done.

In addition, the application must call :c:func:`os_mgmt_register_group()` and :c:func:`img_mgmt_register_group()` to register the image and OS management command handlers.
The application must also call :c:func:`smp_bt_register()` to register the SMP Bluetooth Service.

See the :file:`thingy53.c` source file in a compatible sample for how this is done.

.. _thingy53_app_external_flash:

External flash
==============

During a FOTA update, there might not be enough available flash storage to store the existing application and network core images, so the incoming images must be stored in external flash storage.
This means that a sample compatible with the Nordic Thingy:53 must enable the external flash using the QSPI driver.

> To do this, e by enabling :kconfig:`CONFIG_NORDIC_QSPI_NOR`, and setting :kconfig:`CONFIG_NORDIC_QSPI_NOR_FLASH_LAYOUT_PAGE_SIZE` to ``4096`` and :kconfig:`CONFIG_NORDIC_QSPI_NOR_STACK_WRITE_BUFFER_SIZE` to ``16``.

Reviewer 5: "To do this, e by enabling.." <-- it looks like something happened to the second part of this sentence.

Author: This sentence used to be part of the previous sentence, and apparently my conversion work was interrupted. Fixed.

Thingy:53 application guide
***************************

The Nordic Thingy:53 does not have a built-in J-Link debug IC.

> Because of that, the samples and applications that are compatible with the Nordic Thingy:53 by default use MCUboot bootloader with serial recovery support.

Reviewer 4: Suggest rewording this sentence to: _Because of that, the samples and applications that are compatible with the Nordic Thingy:53 by default include the MCUboot bootloader with serial recovery support._

Include is in my opinion more descriptive as the MCUbootloader is a a separate image that is merged with the application.

Author: Applied.

> Applications compatible with the Nordic Thingy:53 can be updated using DFU functionality with either `nRF Connect Programmer`_ or the `nRF Programmer`_ mobile application.

Reviewer 3: ```suggestion
You can update applications compatible with the Nordic Thingy:53 using the DFU functionality with either `nRF Connect Programmer`_ or the `nRF Programmer`_ mobile application.
```

Author: Fixed.

Enabling DFU and MCUboot requires consistency in configuration of the samples and applications.
For a sample or application to be compatible with Thingy:53, the application must comply with the configurations described in the following sections.

.. note::
   To use your application with the Nordic Thingy:53 preprogrammed bootloader, you must only set the proper configuration for the Partition Manager and MCUboot.

>    Other mentioned features are nice to have, but they are not mandatory.

Reviewer 4: Suggest rephrasing line 291 to: Configurations like USB logging and FOTA are not mandatory, but can be enabled if needed.

Author: Adapted to not mention specific features for increased compatibility and future-proofing.

>    All of applications compatible with the Nordic Thingy:53 in the |NCS| also support the mentioned additional features.

Reviewer 3: ```suggestion
   All applications compatible with the Nordic Thingy:53 in the |NCS| also support the additional features.
```

Author: Fixed.

See :file:`pm_static_thingy53_nrf5340_cpuapp.yml` in a compatible sample for the configuration used.
The memory layout must stay consistent, so that MCUboot can perform proper image updates and clean up the settings storage partition.

The PCD SRAM partition is locked by the MCUboot bootloader to prevent the application from modifying the network core firmware.
Trying to access data on this partition results in an ARM fault.

It is also necessary to add a flash controller overlay for the application core to allow data to be loaded from the external flash memory to the network core using the application core RAM.
This is done because the network core cannot access the external flash memory directly.
See :file:`thingy53_nrf5340_cpuapp.overlay` in a compatible sample for how this is done.

.. _thingy53_app_mcuboot_bootloader:

MCUboot bootloader
==================

Each sample must enable the MCUboot bootloader and use the same configuration of the bootloader as all of the samples in the |NCS| that are compatible with the Nordic Thingy:53.

> This is needed to let MCUboot use the sample as an image update.

Reviewer 4: Suggest rephrasing to: Each sample must enable the MCUboot bootloader and use the same configuration of the bootloader as all the Nordic Thingy:53 compatible samples in the |NCS|.
This is ensures that the sample includes the MCUboot bootloader and that a MCUboot compatible image is generated when the sample is built.

Author: Applied.

The MCUboot bootloader supports serial recovery and a custom command to erase the settings storage partition.
Erasing the settings partition is needed to ensure that an application is not booted with incompatible content loaded from the settings partition.

> The MCUboot bootloader is enabled by using the following Kconfigs:

Reviewer 3: ```suggestion
To enable the MCUboot bootloader use the following Kconfig options:
```

Author: Fixed.

* :kconfig:`CONFIG_BOOTLOADER_MCUBOOT`
* :kconfig:`CONFIG_IMG_MANAGER`
* :kconfig:`CONFIG_MCUBOOT_IMG_MANAGER`
* :kconfig:`CONFIG_IMG_ERASE_PROGRESSIVELY`

> In addition, :kconfig:`CONFIG_UPDATEABLE_IMAGE_NUMBER` must be set to ``2`` and :kconfig:`CONFIG_MCUBOOT_IMAGE_VERSION` must be set to ``"1.0.0+0"``.

Reviewer 3: ```suggestion
In addition, set the :kconfig:`CONFIG_UPDATEABLE_IMAGE_NUMBER` option to ``2`` and :kconfig:`CONFIG_MCUBOOT_IMAGE_VERSION` to ``"1.0.0+0"``.
```

Author: Fixed.

See :file:`thingy53_nrf5430_cpuapp.conf` in a compatible sample for how this is done.

.. _thingy53_app_usb:

USB
===

An application compatible with the Nordic Thingy:53 is expected to enable USB CDC ACM as back-end for logging.
The logs are provided using USB CDC ACM to allow accessing them without additional hardware.

Most of the applications and samples compatible with the Nordic Thingy:53 use only a single instance of USB CDC ACM that works as the logger's back-end.
No other USB classes are used.
These samples can share a common USB product name, vendor ID, and product ID.
If a sample supports additional USB classes or more than one instance of USB CDC ACM, it must use a dedicated product name, vendor ID, and product ID.

> The USB device stack and the CDC ACM class are enabled by using :kconfig:`CONFIG_USB_DEVICE_STACK` and :kconfig:`CONFIG_USB_CDC_ACM`.

Reviewer 3: ```suggestion
To enable the USB device stack and the CDC ACM class use the options :kconfig:`CONFIG_USB_DEVICE_STACK` and :kconfig:`CONFIG_USB_CDC_ACM`.
```

Author: Fixed.

Author: Combined this and the next three suggestions into a bullet point list to improve flow.

> The default log level should be set to Error.

Reviewer 3: ```suggestion
Set the default log level to Error.
```

> This is done by enabling :kconfig:`CONFIG_USB_CDC_ACM_LOG_LEVEL_ERR` and :kconfig:`CONFIG_USB_DEVICE_LOG_LEVEL_ERR`.

Reviewer 3: ```suggestion
Enable the options :kconfig:`CONFIG_USB_CDC_ACM_LOG_LEVEL_ERR` and :kconfig:`CONFIG_USB_DEVICE_LOG_LEVEL_ERR`.
```

> Also, the UART log back-end must be enabled by using :kconfig:`CONFIG_LOG_BACKEND_UART`, and SEGGER RTT must be disabled by using both :kconfig:`CONFIG_LOG_BACKEND_RTT` and :kconfig:`CONFIG_LOG_BACKEND_RTT`.

Reviewer 3: ```suggestion
You also must enable the UART log back-end using the :kconfig:`CONFIG_LOG_BACKEND_UART` option and disable SEGGER RTT using both :kconfig:`CONFIG_LOG_BACKEND_RTT` and :kconfig:`CONFIG_LOG_BACKEND_RTT`.
```

See :file:`thingy53_nrf5340cpuapp.conf` in a compatible sample for how this is done.

> In addition, the USB stack must be initialized by calling :c:func:`usb_enable()` from your application.

Reviewer 3: ```suggestion
In addition, you have to call the :c:func:`usb_enable()` function from your application to initialize the USB stack.
```

Author: Fixed.

See the :file:`thingy53.c` source file in a compatible sample for how this is done.

.. _thingy53_app_antenna:

Antenna selection
=================

The Nordic Thingy:53 has an RF front-end with two 2.4 GHz antennas:

* **ANT1** is connected to the nRF5340 through the nRF21540 RF FEM and supports TX gain of up to +20 dBm.
* **ANT2** is connected to the nRF5340 through the RF switch and supports TX output power of up to +3 dBm.

.. figure:: images/thingy53_antenna_connections.svg
   :alt: Nordic Thingy:53 - Antenna connections

   Nordic Thingy:53 - Antenna connections

The samples in the |NCS| that are compatible with the Nordic Thingy:53 use **ANT1** by default, with the nRF21540 gain set to +10 dBM.
The TX gain can be configured with the :kconfig:option:`CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB` Kconfig option to select between +10 dBM or +20 dBM gain.

> To use the **ANT2** antenna, disable the :kconfig:option:`CONFIG_MPSL_FEM` Kconfig option.

Reviewer 2: Hi @bjornspock
Sorry for late reply. It seems that when we disable the `CONFIG_MPSL_FEM`, the pin that is used to control FEM's PDN and switch to using ANT2 is not controlled at all (it has neither pull up nor pull down then).

> To use the **ANT2** antenna, disable the :kconfig:option:`CONFIG_MPSL_FEM` Kconfig option.

Reviewer 2: Hi @bjornspock
I confirmed with @ankuns that the FEM pins are not controlled when you disable the `CONFIG_MPSL_FEM`. In that case controlling pins is board or user's responsibility. It seems that with the current implementation the pins are floating (the board does not set them) which may lead to troubles.

.. note::

>    Transmitting wth TX output power above +10 dBM is not permitted in some regions.

Reviewer 3: ```suggestion
   Transmitting with TX output power above +10 dBM is not permitted in some regions.
```

Author: Fixed.

>    See the `Nordic Thingy:53 Hardware`_ documentation for the applicable regulations in your region before changing the configuration.

Author: Applied.

.. _thingy53_app_fota_smp:

FOTA and SMP
============

FOTA updates on the Nordic Thingy:53 are supported using MCUmanager's Simple Management Protocol (SMP) over Bluetooth.
The application acts as a GATT server and allows the connected Bluetooth Central to perform the firmware update for both the application and network core.
The Bluetooth configuration is updated to allow quick DFU data trasfer.

The application supports SMP handlers related to:

* Image management.
* Operating System (OS) management -- used to reboot the device after the firmware upload is complete.
* Erasing settings partition -- used to ensure that a new application is not booted with incompatible content in the settings partition written by the previous application.

> To enable the MCUmanager library and handling of SMP commands over Bluetooth, use :kconfig:`CONFIG_MCUMGR` and :kconfig:`CONFIG_MCUMGR_SMP_BT`.

Reviewer 3: ```suggestion
To enable the MCUmanager library and handling of SMP commands over Bluetooth, use the options :kconfig:`CONFIG_MCUMGR` and :kconfig:`CONFIG_MCUMGR_SMP_BT`.
```

Author: Fixed.

Also, disable encryption and authentication requirement for the Bluetooth SMP transport using :kconfig:`CONFIG_MCUMGR_SMP_BT_AUTHEN`.

> The MCUmanager handlers for image and OS management must be enabled.

Author: Fixed.

> To do this, use :kconfig:`CONFIG_MCUMGR_CMD_IMG_MGMT` and :kconfig:`CONFIG_MCUMGR_CMD_OS_MGMT`.

Reviewer 3: ```suggestion
Use the options :kconfig:`CONFIG_MCUMGR_CMD_IMG_MGMT` and :kconfig:`CONFIG_MCUMGR_CMD_OS_MGMT`.
```

Author: Fixed.

> The MCUmanager processing of basic Zephyr group commands and storage erase commands must also be enabled.

Author: Fixed.

> This is done using both :kconfig:`CONFIG_MCUMGR_GRP_ZEPHYR_BASIC` and :kconfig:`CONFIG_MCUMGR_GRP_BASIC_CMD_STORAGE_ERASE`.

Reviewer 3: ```suggestion
Use the options :kconfig:`CONFIG_MCUMGR_GRP_ZEPHYR_BASIC` and :kconfig:`CONFIG_MCUMGR_GRP_BASIC_CMD_STORAGE_ERASE`.
```

Author: Fixed.

See :file:`thing53_nrf5340_cpuapp.conf` in a compatible sample for how this is done.

> In addition, the application must call :c:func:`os_mgmt_register_group()` and :c:func:`img_mgmt_register_group()` to register the image and OS management command handlers.

Author: Fixed.

The application must also call :c:func:`smp_bt_register()` to register the SMP Bluetooth Service.

See the :file:`thingy53.c` source file in a compatible sample for how this is done.

.. _thingy53_app_external_flash:

External flash
==============

During a FOTA update, there might not be enough available flash storage to store the existing application and network core images, so the incoming images must be stored in external flash storage.
This means that a sample compatible with the Nordic Thingy:53 must enable the external flash using the QSPI driver.

> To do this, enable :kconfig:`CONFIG_NORDIC_QSPI_NOR`, and set :kconfig:`CONFIG_NORDIC_QSPI_NOR_FLASH_LAYOUT_PAGE_SIZE` to ``4096`` and :kconfig:`CONFIG_NORDIC_QSPI_NOR_STACK_WRITE_BUFFER_SIZE` to ``16``.

Author: Fixed.

To ensure correct operation of the flash simulator, you must also enable :kconfig:option:`CONFIG_FLASH_NOP_DEVICE`.

See :file:`thingy53_nrf5340_cpuapp.conf` in a compatible sample for how this is done.

.. _thingy53_compatible_applications:

Samples and applications compatible with Thingy:53

> ==================================================

Author: Added in a different PR, pulling that in with a rebase.
